crates/rpc/rpc/src/eth/gas_oracle.rs

```diff
@@ -0,0 +1,264 @@
+pub const DEFAULT_IGNORE_PRICE: U256 = U256::from_limbs([2u64, 0, 0, 0]);
+
+/// Settings for the [GasPriceOracle]
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct GasPriceOracleConfig {
+    /// The number of populated blocks to produce the gas price estimate
+    pub blocks: u32,
+
+    /// The percentile of gas prices to use for the estimate
+    pub percentile: u32,
+
+    /// The maximum number of headers to keep in the cache
+    pub max_header_history: u64,
+
+    /// The maximum number of blocks for estimating gas price
+    pub max_block_history: u64,
+
+    /// The default gas price to use if there are no blocks to use
+    pub default: Option<U256>,
+
+    /// The maximum gas price to use for the estimate
+    pub max_price: Option<U256>,
+
+    /// The minimum gas price, under which the sample will be ignored
+    pub ignore_price: Option<U256>,
+}
+
+impl Default for GasPriceOracleConfig {
+    fn default() -> Self {
+        GasPriceOracleConfig {
+            blocks: 20,
+            percentile: 60,
+            max_header_history: 1024,
+            max_block_history: 1024,
+            default: None,
+            max_price: Some(DEFAULT_MAX_PRICE),
+            ignore_price: Some(DEFAULT_IGNORE_PRICE),
+        }
+    }
+}
+
+/// Calculates a gas price depending on recent blocks.
+#[derive(Debug)]
+pub struct GasPriceOracle<Client> {
+    /// The type used to subscribe to block events and get block info
+    client: Client,
+    /// The cache for blocks
+    cache: EthStateCache,
+    /// The config for the oracle
+    oracle_config: GasPriceOracleConfig,
+    /// The latest calculated price and its block hash
+    last_price: Mutex<GasPriceOracleResult>,
+}
+
+impl<Client> GasPriceOracle<Client>
+where
+    Client: BlockProviderIdExt + 'static,
+{
+    /// Creates and returns the [GasPriceOracle].
+    pub fn new(
+        client: Client,
+        mut oracle_config: GasPriceOracleConfig,
+        cache: EthStateCache,
+    ) -> Self {
+        // sanitize the perentile to be less than 100
+        if oracle_config.percentile > 100 {
+            warn!(prev_percentile=?oracle_config.percentile, "Invalid configured gas price percentile, using 100 instead");
+            oracle_config.percentile = 100;
+        }
+
+        Self { client, oracle_config, last_price: Default::default(), cache }
+    }
+
+    /// Suggests a gas price estimate based on recent blocks, using the configured percentile.
+    pub async fn suggest_tip_cap(&self) -> EthResult<U256> {
+        let header_hash = self
+            .client
+            .block_hash_for_id(BlockId::Number(BlockNumberOrTag::Latest))?
+            .expect("a latest header should always exist");
+        let header = self
+            .client
+            .header_by_hash_or_number(BlockHashOrNumber::Hash(header_hash))?
+            .ok_or(GeneralError::Provider(ProviderError::BlockHash { block_hash: header_hash }))?;
```

Comment:
we have UnknownBlockNumber for these cases. 
let's replace the except and the `BlockHash` error with that.

https://github.com/paradigmxyz/reth/blob/1776a0c8e0da24532a7d3ccca0ac1e189da344ee/crates/rpc/rpc/src/eth/error.rs#L26-L27